pkg/ingester/activeseries/active_series.go

```diff
@@ -200,34 +225,77 @@ func (s *seriesStripe) updateSeriesTimestamp(now time.Time, series labels.Labels
 	}
 }
 
-func (s *seriesStripe) findEntryForSeries(ref uint64) *atomic.Int64 {
+func (s *seriesStripe) findEntryForSeries(ref uint64, hasNativeHistograms bool, numNativeHistogramBuckets int) (*atomic.Int64, bool) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.refs[ref].nanos
+	entry := s.refs[ref]
+	return entry.nanos, entry.isNativeHistogram != hasNativeHistograms || entry.numNativeHistogramBuckets != numNativeHistogramBuckets
 }
 
-func (s *seriesStripe) findOrCreateEntryForSeries(ref uint64, series labels.Labels, nowNanos int64) (*atomic.Int64, bool) {
+func (s *seriesStripe) findAndUpdateOrCreateEntryForSeries(ref uint64, series labels.Labels, nowNanos int64, hasNativeHistograms bool, numNativeHistogramBuckets int) (*atomic.Int64, bool) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	matches := s.matchers.matches(series)
```

Comment:
This is a linear algorithm that only depends on the labels of the series, not whether it's float or native histogram. This means that it should not be executed if ref already exists. Let's move it back to its original place. In the `if ok` part you can simply get the matchers from the entry, as in
```
if ok {
  matches := entry.matches
  matchesLen := matches.len()
```

Reply:
Hmm I don't think this line has been moved in this branch. `matches` is used both in the `if ok` block and afterwards, so it makes sense to leave it here rather than create it in 2 separate places.